jax_galsim/interpolatedimage.py

```diff
@@ -0,0 +1,1194 @@
+    location to interpolate and then over the nterpolation kernel.
+
+    In JAX, we instead write things such that the loop over the points to be interpolated
+    is vectorized in the code. We represent the loops over the interpolation kernel as explicit
+    for loops.
+    """
+    # the vectorization over the interpolation points is easier to think about
+    # if they are in a 1D array. So we use ravel to flatten them and then reshape
+    # at the end.
+    orig_shape = x.shape
+
+    # the variables here are
+    #  x/y: the x/y coordinates of the points to be interpolated
+    #  xi/yi: the index of the nerest pixel below the point
+    #  xp/yp: the x/y coordinate of the nearest pixel below the point
+    #  nx/ny: the size of the x/y arrays
+    x = x.ravel()
+    xi = jnp.floor(x - xmin).astype(jnp.int32)
+    xp = xi + xmin
+    nx = zp.shape[1]
+
+    y = y.ravel()
+    yi = jnp.floor(y - ymin).astype(jnp.int32)
+    yp = yi + ymin
+    ny = zp.shape[0]
+
+    # this function is the inner loop over the x direction
+    # the variables are
+    #  i: the index of the location in the interpolation kernel
+    #  z: the final interpolated values
+    #  wy: the weight of the interpolation kernel in the y direction
+    #  msky: a mask that is true if the y index is in bounds
+    #  yind: the y index of the interpolation point needed by the kernel
+    def _body_1d(i, args):
+        z, wy, msky, yind, xi, xp, zp = args
+
+        # this block computes the x weight using the
+        # offset in the interpolation kernel i
+        xind = xi + i
+        mskx = (xind >= 0) & (xind < nx)
+        _x = x - (xp + i)
+        wx = interp._xval_noraise(_x)
+
+        # the actual interpolation is done here.
+        # we use jnp.where to only do the interpolation
+        # where the x and y indices are in bounds.
+        # the total weight is the product of the x and y weights.
+        w = wx * wy
+        msk = msky & mskx
+        z += jnp.where(msk, zp[yind, xind] * w, 0)
+
+        return [z, wy, msky, yind, xi, xp, zp]
+
+    # this function is the outer loop over the y direction
+    # the variables are
+    #  i: the index of the location in the interpolation kernel
+    #  z: the final interpolated values
+    def _body(i, args):
+        z, xi, yi, xp, yp, zp = args
+
+        # this block computes the x weight using the
+        # offset in the interpolation kernel i
+        yind = yi + i
+        msk = (yind >= 0) & (yind < ny)
+        _y = y - (yp + i)
+        wy = interp._xval_noraise(_y)
+
+        # this call computes the interpolant for each x locatoon that gets
+        # paired with this y location
+        z = jax.lax.fori_loop(
+            -interp.xrange, interp.xrange + 1, _body_1d, [z, wy, msk, yind, xi, xp, zp]
+        )[0]
+        return [z, xi, yi, xp, yp, zp]
+
+    # the actual loop call for y is here
+    z = jax.lax.fori_loop(
+        -interp.xrange,
+        interp.xrange + 1,
```

Comment:
are these supposed to be `yrange` ? or always the same as `xrange`

Reply:
Name is correct. The interpolants are all 1D and use x as the variable.